osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs

```diff
@@ -14,51 +16,115 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+                    aimStrain += angleBonus;
+                }
+
+                if (prevVector.Length > currVector.Length)
                 {
-                    const double scale = 90;
+                    double velChangeBonus = Math.Max(0, Math.Sqrt((prevVector.Length - currVector.Length) * currVector.Length) - currVector.Length) * Math.Min(1, osuCurrObj.JumpDistance / 100);
 
-                    var angleBonus = Math.Sqrt(
-                        Math.Max(osuPrevious.JumpDistance - scale, 0)
-                        * Math.Pow(Math.Sin(osuCurrent.Angle.Value - angle_bonus_begin), 2)
-                        * Math.Max(osuCurrent.JumpDistance - scale, 0));
-                    result = 1.4 * applyDiminishingExp(Math.Max(0, angleBonus)) / Math.Max(timing_threshold, osuPrevious.StrainTime);
+                    aimStrain += velChangeBonus * vel_change_multiplier * Math.Sqrt(100 / osuCurrObj.StrainTime);
                 }
             }
+            else // There is a rhythm change
+            {
+                // Rewarding rhythm, take the smaller velocity as base.
+                double rhythmBonus = Math.Min(currVector.Length, prevVector.Length);
 
-            double jumpDistanceExp = applyDiminishingExp(osuCurrent.JumpDistance);
-            double travelDistanceExp = applyDiminishingExp(osuCurrent.TravelDistance);
+                if (osuCurrObj.StrainTime + 10 < osuPrevObj.StrainTime && osuPrevObj.StrainTime > osuLastObj.StrainTime + 10)
+                    // Don't want to reward for a rhythm change back to back (unless its a double, which is why this only checks for fast -> slow -> fast).
+                    rhythmBonus = 0;
 
-            return Math.Max(
-                result + (jumpDistanceExp + travelDistanceExp + Math.Sqrt(travelDistanceExp * jumpDistanceExp)) / Math.Max(osuCurrent.StrainTime, timing_threshold),
-                (Math.Sqrt(travelDistanceExp * jumpDistanceExp) + jumpDistanceExp + travelDistanceExp) / osuCurrent.StrainTime
-            );
+                aimStrain += rhythmBonus * rhythm_variance_multiplier; // add in rhythm velocity.
+            }
+
+            if (osuCurrObj.TravelDistance != 0)
+            {
+                double sliderBonus = Math.Max(osuCurrObj.TravelDistance, slider_jump_multiplier * Math.Sqrt(osuCurrObj.TravelDistance * osuCurrObj.JumpDistance)) / osuCurrObj.StrainTime;
+
+                // Add in slider velocity.
+                aimStrain += sliderBonus * slider_multiplier;
+            }
+
+            return aimStrain;
         }
 
-        private double applyDiminishingExp(double val) => Math.Pow(val, 0.99);
+        private double calcWideAngleBonus(double angle)
+        {
+            if (angle < Math.PI / 3)
+                return 0;
+            if (angle < 2 * Math.PI / 3)
+                return Math.Pow(Math.Sin(1.5 * (angle - Math.PI / 3)), 2);
+
+            return 0.25 + 0.75 * Math.Pow(Math.Sin(1.5 * (Math.PI / 3 - (angle - 2 * Math.PI / 3))), 2);
```

Comment:
`Math.PI / 3 - (angle - 2 * Math.PI / 3))` is a pretty convoluted way of writing `Math.PI - angle`. was this the intention here?

Reply:
Yeah this is pretty dumb. I was visualizing the angle change weirdly. As long as the graph is the same the change is fine.

Reply:
Actually now that you talk about it, on comparing the desmos doesn't even have the `2pi / 3` factor in. So I'm just going to remove that to match.